src/main/scala/is/hail/expr/ir/functions/UtilFunctions.scala

```diff
@@ -17,23 +17,92 @@ object UtilFunctions extends RegistryFunctions {
       ArrayFold(a, zero, "sum", "v", If(IsNA(Ref("v")), Ref("sum"), ApplyBinaryPrimOp(Add(), Ref("sum"), Ref("v"))))
     }
 
+    registerIR("product", TArray(tnum("T"))) { a =>
+      val product = genUID()
+      val v = genUID()
+      val one = Literal(1, coerce[TArray](a.typ).elementType)
+      ArrayFold(a, one, product, v, If(IsNA(Ref(v)), Ref(product), ApplyBinaryPrimOp(Multiply(), Ref(product), Ref(v))))
+    }
+
     registerIR("sum", TAggregable(tnum("T")))(ApplyAggOp(_, Sum(), Seq()))
 
+    registerIR("count", TAggregable(tv("T"))) { agg =>
+      val uid = genUID()
+      ApplyAggOp(AggMap(agg, uid, MakeStruct(Seq())), Count(), Seq())
+    }
+
     registerIR("min", TArray(tnum("T"))) { a =>
-      val body = If(IsNA(Ref("min")),
-        Ref("value"),
-        If(IsNA(Ref("value")),
-          Ref("min"),
-          If(ApplyBinaryPrimOp(LT(), Ref("min"), Ref("value")), Ref("min"), Ref("value"))))
-      ArrayFold(a, NA(tnum("T").t), "min", "value", body)
+      val min = genUID()
+      val value = genUID()
+      val body = If(IsNA(Ref(min)),
+        Ref(value),
+        If(IsNA(Ref(value)),
+          Ref(min),
+          If(ApplyBinaryPrimOp(LT(), Ref(value), Ref(min)), Ref(value), Ref(min))))
+      ArrayFold(a, NA(tnum("T").t), min, value, body)
+    }
+
+    registerIR("max", TArray(tnum("T"))) { a =>
+      val max = genUID()
+      val value = genUID()
+      val body = If(IsNA(Ref(max)),
+        Ref(value),
+        If(IsNA(Ref(value)),
+          Ref(max),
+          If(ApplyBinaryPrimOp(GT(), Ref(value), Ref(max)), Ref(value), Ref(max))))
+      ArrayFold(a, NA(tnum("T").t), max, value, body)
     }
 
     registerIR("isDefined", tv("T")) { a => ApplyUnaryPrimOp(Bang(), IsNA(a)) }
+    registerIR("isMissing", tv("T")) { a => IsNA(a) }
 
     registerIR("[]", TArray(tv("T")), TInt32()) { (a, i) => ArrayRef(a, i) }
 
+    registerIR("[:]", TArray(tv("T"))) { (a) => a }
+
+    registerIR("[*:]", TArray(tv("T")), TInt32()) { (a, i) =>
```

Comment:
min/max/slicing all look great. Worth adding simple tests to CompileSuite?

Reply:
Hmm.  Can I get back to you on a strategy for this?  We're in an awkward point where we don't want a completely duplicated set of tests for expressions and the tests that test AST will eventually test IR when everything gets moved over.

Reply:
Yup